qiskit/circuit/controlflow/for_loop.py

```
@@ -201,6 +270,7 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         # Loops do not need to pass any further resources in, because this scope itself defines the
         # extent of ``break`` and ``continue`` statements.
         body = scope.build(scope.qubits(), scope.clbits())
+        body.calibrations = self._circuit.calibrations
```

Sharing a pointer to the calibrations data in loads of places is _really_ likely to cause it to get out-of-sync in weird ways.  Besides this, the calibrations dictionary doesn't mean the same thing in circuit scopes; the relative qubit indices are stored in the dictionary as part of the lookup, and those relative indices (in general) change when you go into a control-flow block.

If your concern is lookup of calibrations, I'd say that the answer for look-up is that things that both recursive into control-flow ops and care about calibrations should be passing down the top-level calibrations dictionary and handling it themselves.

If your concern is about causing `QuantumCircuit.add_calibration` to work while in a builder block, I'd mostly just suggest that the method should raise an exception if you attempt to set a "scoped" calibration.  Calibrations are global and we shouldn't hide that.

---

I thought the qubit indices present in the calibrations represented the physical qubits on which, after transpilation with a trivial layout pass, the calibrations would be considered available. The concern was about the fact that when passing through a transpiler pass, the basis translator would wrongly decompose the gates within the body of the control flow ops despite a calibration having been introduced at the wider level. That is why I thought of this referencing. 

So I definitely care more about your point 1, so would you suggest I fix this in the transpiler directly?